### PR TITLE
Add a range field and support steps in text fields

### PR DESCRIPTION
**Pull request by fritzmg**
This PR adds a `range` input field to the form generator.

It also allows the `step` value to be set for `number` input fields.

I still used the old method of re-using the `minlength` and `maxlength` fields - so it still needs to be adjusted when https://github.com/contao/contao/issues/437 gets implemented.

### CODE REVIEW

**Review by aschempp — DISMISSED** [review]

**Comment by Toflar** [comment]
I've added the `[WIP]` prefix so the PR does not get merged in it's current state as it depends on #437.

**Comment by fritzmg** [comment]
I have updated the PR using `minval` and `maxval`. Should I rebase this to `4.8`?

**Comment by leofeyer** [comment]
No, Contao 4.8 is feature complete, so you can leave the branch as it is.

**Comment by fritzmg** [comment]
Oh that's too bad. I only made the changes just now, because the depending change for 4.8 was made just now: https://github.com/contao/contao/issues/437#issuecomment-511863261

**Comment by leofeyer** [comment]
@fritzmg This PR needs to be rebased onto the current master branch so we can review it. 🙈

**Comment by fritzmg** [comment]
Rebased and ready to review.

File: core-bundle/src/Resources/contao/forms/FormRange.php [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+						$this->strId,
+						($this->hideInput ? ' password' : ''),
+						(('' !== $this->strClass) ? ' '.$this->strClass : ''),
+						StringUtil::specialchars($this->value),
+						$this->getAttributes(),
+						$this->strTagEnding);
+	}
+}
+
+class_alias(FormRange::class, 'FormRange');
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/forms/FormRange.php]: Imho that's not needed. It's a new class that's being introduced and it doesn't need to be aliased into the global namespace.

**Review by Toflar — COMMENTED** [review]
One minor detail, LGTM otherwise 👍

**Review by ausi — DISMISSED** [review]

**Review by aschempp — DISMISSED** [review]

File: core-bundle/src/Resources/contao/forms/FormRange.php (hunk 1 above)
Comment (fritzmg) [reply core-bundle/src/Resources/contao/forms/FormRange.php]: Removed now :)

**Review by Toflar — DISMISSED** [review]

**Comment by leofeyer** [comment]
Contao 4.9 is now in the review phase in which we try to review and merge the remaining pull requests that have been created during the development phase. Your pull request does not yet meet the requirements to be reviewed though.

* Make sure that the PR is feature complete.
* Rebase the PR onto the latest master branch.
* Address the requested changes if any.
* Make sure that all CI checks are successful.

Please fix this soon if you can, because the review phase ends on January 15th and we need some time to review the PR before we can merge it.

**Comment by fritzmg** [comment]
This PR is rebased and ready - but I am unsure on how to properly deal with the remaining code style fail. I ran `vendor/bin/php-cs-fixer fix`, however it did not change `FormRange.php` according to the coding style check. I can of course change it manually to what the check is showing me on GitHub, but I wonder why the CS fixer is not actually fixing it for me ...

**Comment by leofeyer** [comment]
You can try `vendor/bin/php-cs-fixer fix -v --using-cache=no`.

**Comment by fritzmg** [comment]
Got it now, I had to use the following command, so that the `legacy` configuration of the CS fixer is used:

```sh
PHP_CS_CONFIG="legacy" vendor/bin/php-cs-fixer fix
```

**Comment by leofeyer** [comment]
Just run `composer run php-cs-fixer` and stop worrying about the details. 😉

**Comment by fritzmg** [comment]
Thank you, forgot about `composer run …` again ;)

**Review by leofeyer — CHANGES_REQUESTED** [review]
While I agree with adding the step attribute to the text field, I fail to see why we need a separate `FormRange` class. Its code is almost identical to the `FormTextField` class and we could easily implement the range type there by adding a new input validation.

However, I find the range field pretty much unusable, as there is no information about which value is currently selected or which range is actually selectable:



How is this intended to be used?

**Comment by fritzmg** [comment]
> While I agree with adding the step attribute to the text field, I fail to see why we need a separate `FormRange` class. Its code is almost identical to the `FormTextField` class and we could easily implement the range type there by adding a new input validation.

Fine with me, I can change the PR accordingly :)

> However, I find the range field pretty much unusable, as there is no information about which value is currently selected or which range is actually selectable:
> 
> 
> 
> How is this intended to be used?

That's up to the application/theme developer, imho.

**Comment by leofeyer** [comment]
> Fine with me, I can change the PR accordingly :)

Thank you.

> That's up to the application/theme developer, imho.

I guess so. 🤷‍♂

**Comment by asaage** [comment]
> > How is this intended to be used?
> 
> That's up to the application/theme developer, imho.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range
[...The precise value, however, is not considered important...]

just stumbled about this... https://jsfiddle.net/w8bw4dte/
not that i favour this but it is a possibility to add to the template.

**Comment by fritzmg** [comment]
👍 I didn't know about the `<output>` element actually. @leofeyer wdyt? Browser support is fairly limited on Microsoft's end though.

**Comment by asaage** [comment]
I think it could be any other element and the question is do we want a **oninput** -handler there.

**Comment by Toflar** [comment]
That's just Javascript. It's not native browser behaviour so I don't think we should support it. We might want to add support for the data-list once it's unified across all browsers though.

**Comment by asaage** [comment]
> We might want to add support for the data-list

But then not seeing what you slide-to gets even worse 🤪

**Comment by leofeyer** [comment]
> That's just Javascript. It's not native browser behaviour so I don't think we should support it.

I am with @Toflar here.

**Comment by fritzmg** [comment]
> > While I agree with adding the step attribute to the text field, I fail to see why we need a separate `FormRange` class. Its code is almost identical to the `FormTextField` class and we could easily implement the range type there by adding a new input validation.
> 
> Fine with me, I can change the PR accordingly :)

Still not sure if this is the right way to go. A new input validation would have the exact same settings as the `digit` validation.

Also setting the form type to "Text field" in order to get an `<input type="range">` seems counter intuitive from a user experience perspective.

**Comment by Toflar** [comment]
I agree, it should be its own widget (might extend or share logic though). Especially if the data-list options become standardized it will be interesting.

**Comment by fritzmg** [comment]
I made a few more changes:

It is perfectly valid to create form fields like this:

```html
<input type="number" min="0.5" max="1.5" step="0.1" value="0.2">
<input type="range" min="0.5" max="1.5" step="0.1" value="0.2">
```

Thus I have changed the `rgxp` of the `minval`, `maxval` and `step` fields accordingly. I also changed the SQL definition of `step` to allow storage of floats. However, I am not sure if `varchar(10)` is the best choice here (that's what `minval` and `maxval` previously used also).

HTML output like this

```html
<input type="range" value="">
```

is invalid, as the content of the `value` attribute must not be empty and it must be numeric, when defined. Thus I also changed the output to cast it to a float, instead of encoding any entities.

**Comment by leofeyer** [comment]
> Also setting the form type to "Text field" in order to get an `<input type="range">` seems counter intuitive from a user experience perspective.

The same is true for `<input type="numeric">`, isn't it? So either we have dedicated classes for all HTML5 input types or we integrate them all into the text field. But please not a mix of both.

**Comment by fritzmg** [comment]
But there is a clear difference between

* `<input type="text">`
* `<input type="text">`
* `<input type="text">`
* `<input type="text">`

**Comment by fritzmg** [comment]
But there is a clear difference between

* `<input type="text">`
* `<input type="text">`

**Comment by fritzmg** [comment]
But there is a clear difference between

* `<input type="date">`
* `<input type="datetime-local">`
* `<input type="email">`
* `<input type="month">`
* `<input type="number">`
* `<input type="password">`
* `<input type="search">`
* `<input type="tel">`
* `<input type="text">`
* `<input type="time">`
* `<input type="url">`
* `<input type="week">`

and

* `<input type="range">`

The former input types all have a text field, where you click and then can enter values manually. The latter is handled completely differently from the others. Sure, some of the input types in the former list have additional client side validation and some added fluff (like spinners, datepickers, etc.). But their concept is similar to one another - while the range slider is a completely different UI concept.

(Sorry for the notification spam, my browser had some issues.)

**Comment by leofeyer** [comment]
Ok, that's a good point. 👍

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Pull request merged**